File: 백준/Silver/2628. 종이자르기/종이자르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		
		st = new StringTokenizer(br.readLine());
		
		int W = Integer.parseInt(st.nextToken());
		int H = Integer.parseInt(st.nextToken());
		
		int N = Integer.parseInt(br.readLine());

		
		List<Integer> cutH = new ArrayList<Integer>();
		List<Integer> cutW = new ArrayList<Integer>();
		cutH.add(0);
		cutW.add(0);
		
		
		int maxW = 0;
		int maxH = 0;
		
		for(int n=0; n<N; n++) {
			st = new StringTokenizer(br.readLine());
			
			int direction = Integer.parseInt(st.nextToken());
		
			//가로로 자를 때 세로 max 결정하기
			if(direction == 0) {
				cutH.add(Integer.parseInt(st.nextToken()));	
				
				
			//세로로 자를 때 가로 max 결정하기
			}else {
				cutW.add(Integer.parseInt(st.nextToken()));

			}
		}
			

			cutW.sort(Comparator.naturalOrder());
			cutH.sort(Comparator.naturalOrder());
			
		
			int temp =0;
			
			for(int i=0; i<cutW.size(); i++) {
				if(i!= cutW.size()-1 ) {
					temp = cutW.get(i+1)-cutW.get(i);
				}
				else if(i==cutW.size()-1) {
					temp = W - cutW.get(i);
				}
			
				if(maxW < temp) maxW = temp;
		
			}
			
			
			for(int i=0; i<cutH.size(); i++) {
				if(i!= cutH.size()-1 ) {
					temp = cutH.get(i+1)-cutH.get(i);
				}
				else if(i==cutH.size()-1) {
					temp = H - cutH.get(i);
				}
			
				if(maxH < temp) maxH = temp;
		
			}
		
			System.out.println(maxH * maxW);
		}

	

}

File: 백준/Silver/2628. 종이자르기/종이자르기.py
w, h = map(int, input().split())
n = int(input())

w_lst = [0,w]
h_lst = [0,h]
for i in range(n):
    way, k = map(int, input().split())
    if way == 0:
        h_lst.append(k)
    else:
        w_lst.append(k)

w_lst.sort()
h_lst.sort()

max_w = 0
max_h = 0
for i in range(1, len(w_lst)):
    if w_lst[i]-w_lst[i-1]>max_w:
        max_w = w_lst[i]-w_lst[i-1]
for j in range(1, len(h_lst)):
    if h_lst[j]-h_lst[j-1]>max_h:
        max_h = h_lst[j]-h_lst[j-1]

print(max_w*max_h)

